### PR TITLE
Add fused DeepSeek V4 FWHT MXFP4 preprocessing

### LINKED ISSUE #800: [Feature]: Add fused DeepSeek V4 FWHT MXFP4 preprocessing
### Problem or use case

DeepSeek V4 indexer inference applies a normalized 128-point Hadamard transform
followed by group-32 MXFP4 quantize/dequantize simulation to both query and
compressed-key tensors.  The published composition is correct but launches a
sequence of framework operations and has no native cuDNN Frontend semantic op.

### Proposed solution

Add an experimental inference-only
`cudnn.ops.fwht_mxfp4_qdq(input_tensor) -> Tensor` operation for contiguous
BF16 tensors whose last dimension is 128.  Keep architecture dispatch, JIT
compilation, launch streams, and kernel scheduling private.  The implementation
is an original CuTe DSL kernel and preserves the official BF16 rounding boundary
and group-32 MXFP4 QDQ semantics exactly.

The initial change includes model-source and Kernel Factory provenance,
bit-exact host references, CTA-boundary tests, the known subnormal
double-rounding regression, FakeTensor/opcheck, `torch.compile(fullgraph=True)`,
non-default-stream ordering, and warmed CUDA Graph replay.

### Why is this important?

This fills a concrete missing primitive for DeepSeek V4 model integration.  On
ComputeLab B200, direct same-process comparisons against the pinned official
composition were bit-exact for all six Q/K model shapes.  The isolated combined
Q+K preprocessing improved by 6.40x, 7.69x, and 8.65x at sequence lengths
8192, 16384, and 32768 respectively.  These are operation-level results, not
full-model speedups.

- [x] I searched existing issues and did not find a duplicate.

### PR DESCRIPTION
**Pull request by YangXu1990uiuc**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I added GitHub labels: one `cat-*`, one or more `mod-*`, and one `orig-*`.
- [ ] I set the **Milestone** and **Projects** fields in the sidebar.  Milestone
  `Frontend 1.29.0` is set; this account lacks the GitHub Projects scope, so a
  maintainer must add the Project field.

## Affected area

FE OSS kernels or CuTeDSL

### Downstream model impact (combined attribution)

On B200, replacing exactly the two preprocessing callsites with executable implementation commit 1addaf4 inside a pinned DeepSeek-V4-Pro ratio-4 Attention forward changed the complete S=4096 learned-indexer -> causal top-k/local-window -> CSA path from **47.922657 ms to 47.779999 ms (1.00299x; 0.2977% latency reduction)**. The current head differs from that measured executable only by review-driven architecture-gate and test hardening. This is forward-only single-layer callsite evidence; its arithmetic 30-layer subtotal is not an assembled model or training result. The isolated preprocessing result of 6.40-8.65x should therefore not be read as downstream model speedup.

- Add experimental `cudnn.ops.fwht_mxfp4_qdq(input_tensor) -> Tensor` for the
  normalized H128 plus group-32 MXFP4 QDQ recipe used by the DeepSeek V4
  indexer.
- Implement the fused operation as an original CuTe DSL kernel with dynamic row
  counts, current-stream execution, per-device compiled-kernel caching, and
  direct `torch.Library` registration.
- Preserve the official BF16 transform boundary and FP32 scale/divide semantics
  exactly, including the subnormal double-rounding regression case.
- Add a pinned model reference, API documentation, exact host
  references, CTA-boundary coverage, opcheck/FakeTensor, `torch.compile`,
  non-default-stream, and warmed CUDA Graph tests.

## Why

DeepSeek V4 applies this preprocessing to both query and compressed-key tensors
before indexer scoring.  The official framework composition has no native cuDNN
Frontend semantic op and launches a sequence of smaller operations.  This change
fills that missing model primitive without exposing architecture, launch, or
schedule details in the public API.

On ComputeLab B200, direct same-process comparison against the pinned official
composition was bit-exact for every element of all six Q/K model shapes.  The
isolated combined Q+K preprocessing improved by 6.40x, 7.69x, and 8.65x at
sequence lengths 8192, 16384, and 32768 respectively.  These are operation-level
results and are not full-model speedup claims.

## Related issues

Fixes #800.

## API and compatibility impact

Adds one experimental, inference-only public operation.  The initial contract is
contiguous, 32-byte-aligned BF16 input with last dimension 128 and a fresh BF16
output of identical shape.  Autograd and implicit layout conversion are rejected.
No existing API behavior changes.

The implementation pins the official DeepSeek source revision. The
in-tree kernel is independently implemented from the documented operation; no external kernel source is included.

## Testing

- `pre-commit run` — passed.
- Host numerical regressions — 4 passed.
- Host direct-Library schema plus public fullgraph FakeTensor smoke — passed.
- ComputeLab job `4012962`, NVIDIA B200 CC 10.0, cuDNN backend 92600,
  PyTorch 2.13.0+cu130:
  `pytest -q test/python/fe_api/dsa/test_DSA_fwht_mxfp4_qdq.py` —
  **23 passed** in 26.80 s.
- The B200 file includes actual kernel cases at M=63/64/65 and 127/128/129,
  exact crafted double-rounding coverage, `torch.library.opcheck`,
  `torch.compile(fullgraph=True)`, gated non-default-stream ordering, and two
  warmed CUDA Graph replays with different inputs.
- Separate model-proxy validation on the same B200 allocation compared the full
  official pinned preprocessing against the public semantic op for Q and K at
  S=8192/16384/32768; all outputs were bit-exact.

## Summary by CodeRabbit

- **New Features**
  - Added an inference-only FWHT/MXFP4 quantize-dequantize operation for supported NVIDIA GPU architectures.
  - Added validation for input type, format, shape, alignment, device, contiguity, and gradient requirements.
  - Added fake-tensor support and stream-aware execution with compilation caching.

- **Documentation**
  - Added the DeepSeek-V4 Indexer Preprocessing API to the API overview.
  - Expanded documentation with implementation details, validation results, and performance benchmarks.

- **Tests**
  - Expanded coverage for operation availability, import failures, input validation, GPU support, and CUDA Graph behavior.

### CODE REVIEW

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost,python_tests

File: docs/fe-oss-apis/deepseek_v4_indexer_preprocess.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+- input and output dtype: BF16
+- input layout: contiguous and 32-byte aligned
+- input values: finite model activations whose FP32 FWHT intermediates remain
+  finite
+- target characterized so far: B200
+- autograd: unsupported
+- framework or layout fallback: none
+
+The model-facing API deliberately contains no `Wrapper`, architecture suffix,
+tile configuration, or explicit compile lifecycle. Its internal compiled
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply docs/fe-oss-apis/deepseek_v4_indexer_preprocess.md:76]: Leaving this unchanged. The cited overview text is a repository-wide APIBase template, while this operation is intentionally a torch.library semantic custom op under cudnn.ops. It owns no user-managed compile, execute, or workspace lifecycle; cudnn.ops.fwht_mxfp4_qdq is the canonical documented surface. Adding nominal Wrapper and Class APIs solely to mirror that stale overview wording would create a duplicate lifecycle that does not exist.

File: python/cudnn/ops/_fwht_mxfp4_qdq_cutedsl.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+        "cvt.rn.satfinite.e2m1x2.bf16x2 quantized, normalized;\n\t"
+        "cvt.rn.scaled::n2::ue8m0.bf16x2.e2m1x2 $0, quantized, $3;\n\t"
+        "}",
+        [
+            U32(values).ir_value(loc=loc, ip=ip),
+            U32(packed_inverse_scale).ir_value(loc=loc, ip=ip),
+            U16(packed_scale_bytes).ir_value(loc=loc, ip=ip),
+        ],
+        "=r,r,r,h",
+    )
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply python/cudnn/ops/_fwht_mxfp4_qdq_cutedsl.py:256]: Fixed in 45756fa86. The launch path now admits only SM100, SM110, and SM120 architecture families, including their minor variants, before importing the CUDA driver or touching the compile cache. A focused test substitutes SM90 and SM130 and asserts that cute.compile is never reached. The complete focused file passed 32 tests on ComputeLab B200.

File: test/python/fe_api/dsa/test_DSA_fwht_mxfp4_qdq.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+    side_stream = torch.cuda.Stream()
+    gate = torch.cuda.Event()
+    with torch.cuda.stream(side_stream):
+        side_stream.wait_event(gate)
+        input_tensor.copy_(replacement)
+        actual = fwht_mxfp4_qdq(input_tensor)
+
+    # A launch incorrectly sent to the default stream would execute before the
+    # gated input update. A launch on the current side stream observes it.
+    gate.record()
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply test/python/fe_api/dsa/test_DSA_fwht_mxfp4_qdq.py]: Fixed in 45756fa86. The test no longer waits on an unrecorded lazy event. It parks the side stream with blocking device work, then enqueues the input replacement and operation on that stream; a wrong default-stream launch deterministically overtakes the replacement and reads the old zeros. The race test is now L1 and GPU-exclusive, and passed in the 32-test focused B200 run.

File: test/python/fe_api/dsa/test_DSA_fwht_mxfp4_qdq.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+        torch.cuda._sleep(500_000_000)
+        input_tensor.copy_(replacement)
+        actual = fwht_mxfp4_qdq(input_tensor)
+
+    torch.cuda.synchronize()
+
+    assert torch.equal(actual, expected)
+
+
+@pytest.mark.L0
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply test/python/fe_api/dsa/test_DSA_fwht_mxfp4_qdq.py]: Fixed in a60634996: the warmed CUDA Graph capture/replay test is now L1. ComputeLab B200 marker collection selected exactly this one L1 case; pre-commit passed.

File: test/python/fe_api/dsa/test_DSA_fwht_mxfp4_qdq.py [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+
+def _require_operation():
+    """Return the operation only when all lazy optional dependencies exist."""
+
+    _require_supported_device()
+    try:
+        import_module("cudnn.ops._fwht_mxfp4_qdq_cutedsl")
+        from cudnn.ops import fwht_mxfp4_qdq
+    except ImportError as error:
+        pytest.skip(f"CuTe DSL optional dependencies are unavailable: {error}")
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply test/python/fe_api/dsa/test_DSA_fwht_mxfp4_qdq.py]: Fixed in a60634996. _require_operation now uses importlib.util.find_spec("cutlass") to skip only when the optional package is genuinely undiscoverable. After that boundary, ImportError from the operation or dependency propagates. Added L0 host contracts for the absent dependency and internal ImportError propagation. ComputeLab B200: 2 passed; pre-commit passed.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost,python_tests

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `a6063499621e` against base `64f6b391144c`.

The semantic API and fused implementation make sense. I found no actionable correctness issue in the checked contracts. The earlier architecture, stream-test and optional-import comments are addressed. The remaining 21–25 us warmed host-call cost is a useful future optimization target for very small inputs, where GPU execution is about 2 us; this does not require adding a user-managed Wrapper lifecycle. One CI coverage correction: the existing frost/python_tests request does not select this fe_api test file. My following standalone `@cudnn-ci-bot run oss` comment will request the relevant test target.

Validation: Matching extension built from a6063499621e; the full focused suite with explicit -m "L0 or L1" passed all 34 cases on SM100 with CuTe DSL 4.8, including exhaustive host numerical checks, subnormal double rounding, CTA tails, opcheck, fullgraph compilation, side-stream ordering, and changed-input/poisoned-output CUDA Graph replay. Additional two-device check: 12 cases across rows 1/17/129 with the ambient device deliberately different from the input device; outputs match bitwise, the ambient device is restored, and each device compiles exactly once. Compared the actual pinned DeepSeek-V4 b5968e9 plus fast-hadamard-transform e7706fa composition against the public FE operation. All int16-view output bits agree at rows 16, 2048 and 524288, including after changing the input and poisoning captured outputs. Source pins and loaded implementation identities were checked. On a 68-SM SM100 device, three ABBA cycles measured warmed CPU call/enqueue time separately from CUDA-event GPU intervals. Each timed graph contains 16 calls to avoid host replay starvation. Rows 16: CPU 37.45 -> 21.56 us, GPU 4.46 -> 1.83 us; rows 2048: CPU 37.70 -> 21.44 us, GPU 5.42 -> 2.08 us; rows 524288: CPU 42.69 -> 24.84 us, GPU 901.38 -> 163.64 us (official -> fused). The isolated merge onto develop 64f6b391144c is conflict-free. Current-head Style passed. Inspected the prior test pipeline: OSS jobs were skipped; the two Python SM100 lanes failed in other operators with DSL compiler errors. These are not validation of the new preprocessing file.

Limitations: The performance checks measure preprocessing only on the listed SM100 workloads; they do not reproduce the B200 model-layer or full-model performance claims. SM110/SM120-family execution was not independently rerun. The same-process official comparison reused the benchmark's guarded TVM-FFI compatibility adapter for TileLang 0.1.8 / apache-tvm-ffi 0.1.8.post2; it handled one empty dataclass-map request and rejected any nonempty mapping. Neither numerical kernel was changed. OSS CI will be pending after the following request; the full target-hardware matrix is not established by these local tests.

Comment only: GitHub does not allow reviewing one's own PR with approval/request-changes.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run oss

**Comment by YangXu1990uiuc** [comment]
**Codex bot review · model `gpt-6-astra`**

CI follow-up for `a6063499621e`: the requested OSS lanes on SM80, SM90 and both SM100 shards passed. Two remaining failures need different handling:

- The guardword scan matches `broken` in the new test file's comment and `broken_internal_import` helper (lines 28, 52 and 55); please rename/reword these.
- The Rubin runner ran out of disk space during setup, before any test executed. That failure provides no kernel result.

These are aggregate lane outcomes; the SM100 lane used CuTe DSL 4.6.2. The 34 focused local checks reported in the review used 4.8 with explicit L0/L1 selection. No duplicate pipeline was started.

**Review by vedaanta — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `a6063499621e` against base `9b9f32781d0c`.

The fused semantic operation is coherent, and I found no approval-blocking correctness issue. The architecture guard, optional-import behavior and revised stream test address the earlier comments. Fresh current-base testing passed all 34 focused cases plus a cache-transition/mixed-exponent probe. Please still address the previously reported test-file guardword failure; that localized CI cleanup is nonblocking under the review policy. This approval does not assert full-model performance or successful Rubin validation.

- **[P2] Still open:** [Reword the test comment and helper rejected by the guardword check](https://github.com/NVIDIA/cudnn-frontend/pull/801#issuecomment-5674387685).

Validation: Reviewed all six changed files and traced the public validation, lazy registration, packed FWHT/QDQ, tail handling, stream selection and per-device cache. A disposable merge of a6063499621e onto develop 9b9f32781d0c was clean; operation and focused-test sources were unchanged by integration. Fresh SM100 integrated run with CuTe DSL 4.7, PyTorch 2.13.0+cu130 and cuDNN 9.26: 34 passed, zero skips, in 13.33 seconds with explicit L0 or L1 selection. Includes exhaustive host BF16/scale checks, actual-kernel subnormal double rounding, CTA tails, opcheck, fullgraph compilation, non-default-stream ordering and changed-input/poisoned-output graph replay. The intended checkout import was verified; a compatible current-base native extension was reused for these Python-only changes. Additional fresh probe: public calls at shapes [128], [17,128], [65537,128], then [128] again matched CPU recipe output bits, preserved input and returned distinct output storage. A 17-row case mixed eight exponent ranges from -125 through 100 across adjacent groups and also matched bitwise. All five calls used exactly one cute.compile invocation, confirming dynamic-row reuse through the small/large/small transition. Verified the scale floor, FP32 divide and BF16 transform boundary against pinned DeepSeek-V4-Pro b5968e9190ef611bbf34a7229255be88a0e937c1 inference/model.py and inference/kernel.py. Reused, rather than reran, the prior same-head review's actual official-composition comparison, two-device checks and ABBA CPU/GPU performance evidence: https://github.com/NVIDIA/cudnn-frontend/pull/801#pullrequestreview-5204950509 . Those preprocessing measurements are not full-model speedups. Current-head Style passed; merge requirements report missing Project. No action-required workflow was found. Pipeline 67893738 still has passing SM80/SM90/two SM100 OSS lanes, the test-file guardword failure, and Rubin job 439954749 failing during setup with no disk space before tests. These aggregate lane outcomes do not replace the fresh focused execution. No duplicate broad CI was launched; the existing run and follow-up already cover this head, with owner cleanup remaining.

Limitations: Fresh GPU execution was SM100 only; SM110/SM120-family operation and Rubin execution were not established here. No new model-level or performance benchmark was run. The added numerical probe uses the reviewed in-tree CPU recipe; actual execution of the pinned external composition is reused prior same-head evidence, not a fresh local run. The focused suite emitted an unregistered xdist_group marker warning and torch deprecation warnings while running serially.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.
